sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCatalog.java

```diff
@@ -0,0 +1,417 @@
+
+/**
+ * Catalog for registering tables and functions. Populates a {@link SimpleCatalog} based on a {@link
+ * SchemaPlus}.
+ */
+public class BeamZetaSqlCatalog {
+  // ZetaSQL function group identifiers. Different function groups may have divergent translation
+  // paths.
+  public static final String PRE_DEFINED_WINDOW_FUNCTIONS = "pre_defined_window_functions";
+  public static final String USER_DEFINED_SQL_FUNCTIONS = "user_defined_functions";
+  public static final String USER_DEFINED_JAVA_SCALAR_FUNCTIONS =
+      "user_defined_java_scalar_functions";
+  /**
+   * Same as {@link Function}.ZETASQL_FUNCTION_GROUP_NAME. Identifies built-in ZetaSQL functions.
+   */
+  public static final String ZETASQL_FUNCTION_GROUP_NAME = "ZetaSQL";
+
+  private static final ImmutableList<String> PRE_DEFINED_WINDOW_FUNCTION_DECLARATIONS =
+      ImmutableList.of(
+          // TODO: support optional function argument (for window_offset).
+          "CREATE FUNCTION TUMBLE(ts TIMESTAMP, window_size STRING) AS (1);",
+          "CREATE FUNCTION TUMBLE_START(window_size STRING) RETURNS TIMESTAMP AS (null);",
+          "CREATE FUNCTION TUMBLE_END(window_size STRING) RETURNS TIMESTAMP AS (null);",
+          "CREATE FUNCTION HOP(ts TIMESTAMP, emit_frequency STRING, window_size STRING) AS (1);",
+          "CREATE FUNCTION HOP_START(emit_frequency STRING, window_size STRING) "
+              + "RETURNS TIMESTAMP AS (null);",
+          "CREATE FUNCTION HOP_END(emit_frequency STRING, window_size STRING) "
+              + "RETURNS TIMESTAMP AS (null);",
+          "CREATE FUNCTION SESSION(ts TIMESTAMP, session_gap STRING) AS (1);",
+          "CREATE FUNCTION SESSION_START(session_gap STRING) RETURNS TIMESTAMP AS (null);",
+          "CREATE FUNCTION SESSION_END(session_gap STRING) RETURNS TIMESTAMP AS (null);");
+
+  /** The top-level Calcite schema, which may contain sub-schemas. */
+  private final SchemaPlus calciteSchema;
+  /**
+   * The top-level ZetaSQL catalog, which may contain nested catalogs for qualified table and
+   * function references.
+   */
+  private final SimpleCatalog zetaSqlCatalog;
+
+  private final JavaTypeFactory typeFactory;
+
+  private final JavaUdfLoader javaUdfLoader = new JavaUdfLoader();
+  private final Map<List<String>, ResolvedNodes.ResolvedCreateFunctionStmt> sqlScalarUdfs =
+      new HashMap<>();
+  /** User-defined table valued functions. */
+  private final Map<List<String>, ResolvedNode> sqlUdtvfs = new HashMap<>();
+
+  private final Map<List<String>, UserFunctionDefinitions.JavaScalarFunction> javaScalarUdfs =
+      new HashMap<>();
+
+  private BeamZetaSqlCatalog(
+      SchemaPlus calciteSchema, SimpleCatalog zetaSqlCatalog, JavaTypeFactory typeFactory) {
+    this.calciteSchema = calciteSchema;
+    this.zetaSqlCatalog = zetaSqlCatalog;
+    this.typeFactory = typeFactory;
+  }
+
+  /** Return catalog pre-populated with builtin functions. */
+  static BeamZetaSqlCatalog create(
+      SchemaPlus topLevelSchema,
```

Comment:
nit: possibly name this `calciteSchema`? It makes it easier to follow if the name doesn't change.

Reply:
done